python/dgl/graphbolt/dataloader.py

```diff
@@ -215,15 +215,38 @@ def __init__(
                     gpu_graph_cache = construct_gpu_graph_cache(
                         sampler, num_gpu_cached_edges, gpu_cache_threshold
                     )
-                datapipe_graph = replace_dp(
-                    datapipe_graph,
-                    sampler,
-                    sampler.fetch_and_sample(
-                        gpu_graph_cache,
-                        get_host_to_device_uva_stream(),
-                        1,
-                    ),
-                )
+                if (
+                    sampler.sampler.__name__ == "sample_layer_neighbors"
+                    or gpu_graph_cache is not None
+                ):
+                    # This code path is not faster for sample_neighbors.
+                    datapipe_graph = replace_dp(
+                        datapipe_graph,
+                        sampler,
+                        sampler.fetch_and_sample(
+                            gpu_graph_cache,
+                            get_host_to_device_uva_stream(),
+                            1,
+                        ),
+                    )
+                elif sampler.sampler.__name__ == "sample_neighbors":
+                    # This code path is faster for sample_neighbors.
+                    datapipe_graph = replace_dp(
+                        datapipe_graph,
+                        sampler,
+                        sampler.datapipe.sample_per_layer(
```

Comment:
If you put the view at the caller of this sample_per_layer. I have no idea what does returning_indices_is_optional to me, since it does not affect the output of this method.

From the caller of sample_per_layer POV, I am setting this bit to turn on a optimization.

Reply:
the simplest solution is renaming it to enable_{awesome_name}_optimization, and document how the optimization works.

Reply:
No user will use sample_per_layer though. It is not exposed anymore. It is only used inside neighbor_sampler.py

Reply:
It has become internal.